### PR TITLE
feat: create `complete-deposit` sign requests

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/563.

## Changes

* Remove some of the unnecessary fields from the `TxSignerEventLoop` and `TxCoordinatorEventLoop`. We can remove the private key field too but it's better to move that into the wallet so I left it as is for now.
* Removed an unnecessary input from the `RotateKeysV1` transaction object.
* Added a method for loading the `SignerWallet` from the database.
* Added trait functions for fetching "fulfilled" requests from the database, partly implement one of these functions for the postgres type.

There is still some outstanding work here:
* ~Rename the `Fulfilled*Request` structs (the requests are only partly fulfilled). Name suggestions are welcome.~ Done.
* ~Fix-up the documentation on the `Fulfilled*Request` structs.~ Done.
* Add an implementation for the in-memory version of the query. Won't do here.
* ~Any other name suggestions are welcome.~ No need, unless they're really really good.

## Testing Information

This PR has tests for the function that loads from the database, but is missing:
1. Properly construct a sign deposit request transaction.
2. Properly fetch nearly fulfilled requests from the database. For now, the query that is here is "sufficient", but we can do it properly after https://github.com/stacks-network/sbtc/issues/585 lands.

Edit: We have some tests for (2) now, but a full suite of tests require https://github.com/stacks-network/sbtc/issues/585 and for us to modify the query included here.

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

File: signer/src/testing/wsts.rs [hunk 1]
```diff
@@ -517,7 +518,7 @@ impl SignerSet {
         &self,
         storage: &S,
         chain_tip: &model::BitcoinBlockHash,
-        aggregate_key: PublicKey,
+        shares: &EncryptedDkgShares,
```
Comment (matteojug) [inline signer/src/testing/wsts.rs:521]: qq: is this just a matter of style, or we gain something by using shares instead of the aggregated key?

File: signer/src/transaction_coordinator.rs [hunk 2]
```diff
@@ -223,32 +230,148 @@ where
...
     }
 
+    /// Transform the finalizable deposit request into a Stacks sign
+    /// request object.
+    /// 
+    /// This function uses bitcoin-core to help with the fee assessment of
+    /// the deposit request, and stacks-core for fee estimation of the
+    /// transaction.
+    #[tracing::instrument(skip_all)]
+    async fn construct_stacks_sign_request(
```
Comment (matteojug) [inline signer/src/transaction_coordinator.rs:308]: nit: rename the fn to include "deposit"?

File: signer/src/stacks/wallet.rs [hunk 3]
```diff
@@ -495,4 +559,82 @@ mod tests {
...
+            block_hash: stacks_chain_tip.block_hash,
+        };
+
+        db.write_stacks_transaction(&tx).await.unwrap();
+        db.write_rotate_keys_transaction(&rotate_keys)
+            .await
+            .unwrap();
+
+        // Okay, now let's load it up and make sure things match.
+        let wallet2 = SignerWallet::load(&ctx, &bitcoin_chain_tip).await.unwrap();
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/wallet.rs:635]: https://github.com/stacks-network/sbtc/pull/615/commits/0ffb0b0471a5b2f9e55fff35cd94d34811aa775b

File: signer/src/storage/model.rs [hunk 4]
```diff
@@ -238,6 +238,86 @@ pub struct Transaction {
...
+    pub sender_script_pub_keys: Vec<ScriptPubKey>,
+}
+
+/// Withdraw request.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, sqlx::FromRow)]
+#[cfg_attr(feature = "testing", derive(fake::Dummy))]
+pub struct FulfilledWithdrawalRequest {
+    /// Transaction ID.
+    pub sweep_txid: BitcoinTxId,
+    /// The transaction fulfillinf the deposit request.
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/model.rs]: https://github.com/stacks-network/sbtc/pull/615/commits/0c88b93eff3bf529edc19e7e98fc36c5b0b2d07f

File: signer/src/storage/model.rs [hunk 5]
```diff
@@ -238,6 +238,86 @@ pub struct Transaction {
...
+    pub sweep_tx: BitcoinTx,
+    /// The block id of the stacks block that includes this transaction
+    pub sweep_block_hash: BitcoinBlockHash,
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/model.rs]: yup, should be bitcoin, fixed in https://github.com/stacks-network/sbtc/pull/615/commits/ab26d27b0aedaf0b1b67014fac2c09563d304d07

File: signer/src/stacks/wallet.rs [hunk 6]
```diff
@@ -145,30 +146,84 @@ impl SignerWallet {
         })
     }
 
+    /// Load the multi-sig wallet from storage.
+    ///
+    /// The wallet that is loaded is the one that cooresponds to the signer
+    /// set defined in the last confirmed key rotation contract callß.
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/wallet.rs]: Fixed in https://github.com/stacks-network/sbtc/pull/615/commits/0ffb0b0471a5b2f9e55fff35cd94d34811aa775b

File: signer/src/testing/wsts.rs (hunk 1 above)
Comment (djordon) [reply signer/src/testing/wsts.rs:521]: Kinda. This is a change in anticipation of another change that will make things a little more clear. I think we should store more data in the database so that the different public keys can be used (I talk about the change in https://github.com/stacks-network/sbtc/issues/614).

**Comment by AshtonStephens** [comment]
Created ticket to track the lack of integration tests here: https://github.com/stacks-network/sbtc/issues/631

File: signer/src/block_observer.rs [hunk 7]
```diff
@@ -288,6 +288,10 @@ where
                 let mut tx_bytes = Vec::new();
                 tx.consensus_encode(&mut tx_bytes)?;
 
+                // TODO: these aren't all sBTC transactions. Some of these
+                // could be "donations". One way to properly label these is
+                // to look at the scriptPubKey of the prevouts of the
+                // transaction.
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/block_observer.rs:294]: Yes, we know we can recognize a donation by it's `scriptPubKey` since they will be locked by something that is known. Specifically, donations are spend-path-only pay-to-taproot UTXOs where the locking public key is the signer aggregate key (the DKG aggregate key). We cannot recognize deposits at all, the signers need to be told that a transaction is a deposit (and then they verify whether that is true).

File: signer/tests/integration/postgres.rs [hunk 8]
```diff
@@ -1362,7 +1364,197 @@ async fn get_signers_script_pubkeys_returns_non_empty_vec_old_rows() {
...
+
+    // This query doesn't *need* bitcoind (it's just a query), we just need
+    // the transaction data in the database. We use the [`TestSweepSetup`]
+    // structure because it has helper functions for generating and storing
+    // sweep transactions, and the [`TestSweepSetup`] structure correctly
+    // sets up the database.
+    let (rpc, faucet) = sbtc::testing::regtest::initialize_blockchain();
+    let setup = TestSweepSetup::new_setup(&rpc, &faucet, 1_000_000, &mut rng);
+
+    // Wwe need to manually update the database with new bitcoin block
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1385]:
```suggestion
    // We need to manually update the database with new bitcoin block
```

**Review by matteojug — COMMENTED** [review]
Other than the mentioned issues in the PR description (and the two missing nits), LGTM.

About `Properly construct a sign deposit request transaction.` mentioned in the testing information, what are you referring to?

File: signer/src/stacks/wallet.rs [hunk 9]
```diff
@@ -495,4 +559,82 @@ mod tests {
...
+            .with_in_memory_storage()
+            .with_mocked_clients()
+            .build();
+        let db = ctx.get_storage_mut();
+
+        // Create a blockchain. We do not generate any withdrawal or
+        // deposit requests, so we do not need to specify the signer set.
+        let test_params = crate::testing::storage::model::Params {
+            num_bitcoin_blocks: 10,
+            num_stacks_blocks_per_bitcoin_block: 0,
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/wallet.rs:587]: Ahh, I believe `(1..0)` is an empty range.

File: signer/tests/integration/postgres.rs (hunk 8 above)
Comment (djordon) [reply signer/tests/integration/postgres.rs]: Done in https://github.com/stacks-network/sbtc/pull/615/commits/97dda8d25209b504926158549605fa0698ba0d3b

File: signer/src/stacks/wallet.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (matteojug) [reply signer/src/stacks/wallet.rs:587]: it seems to be still considering the initial value though:
```
# dbg!((1..0).fold(vec![1], |a, _| a));
[src/main.rs:5:5] (1..0).fold(vec![1], |a, _| a) = [
    1,
]
```

**Review by matteojug — APPROVED** [review]

**Comment by djordon** [comment]
> About `Properly construct a sign deposit request transaction.` mentioned in the testing information, what are you referring to?

I should change the wording here, I want to test that it properly creates a sign deposit request transaction _object_, so that the other signers return a signature for an actual deposit request transaction.

File: signer/src/stacks/wallet.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/wallet.rs:587]: Oh, right, well that settles it. My mistake! 

Also, I did not take in the confusing part about this test, which is that I use Stacks blocks but ask for zero stacks blocks to be generated. I'll fix this in another PR.

> To avoid surprises, should we validate `num_stacks_blocks_per_bitcoin_block > 0` or avoid creating a stacks block if the param is zero?

I would think that we should avoid creating Stacks blocks if we ask for zero. For this code, I should probably just update the number of stacks blocks that are needed to avoid confusion. I'll make sure that it is added in.
